drivers/adodb-mysqli.inc.php

```diff
@@ -1033,13 +1038,176 @@ function Prepare($sql)
+		*/
+		$nQuoteTest = fmod(substr_count($tempSql,"'"),2);
+		
+		if ($nQuoteTest)
+		{
+			$this->outp_throw(
+			"Input string '" . htmlspecialchars($sql) . "' has mismatched single quotes. Quotes inside quoted strings must be correctly escaped",
+			'Execute'
+			);
+			return 0;
+		}				
+		/*
+		* Now break out quoted strings
+		*/
+		preg_match_all('/(\'.*?\')/',$tempSql,$quotedStringArray);
+
+		/*
+		* Do any of the quoted strings contain placeholders?
+		*/
+		$placeholderSearch = preg_grep('/\?/',$quotedStringArray[0]);
+		
+		if (count($placeholderSearch) == 0)
+		{
+			/*
+			* No they don't, we've gone far enough
+			*/
+			return substr_count($sql,'?');
+		}
+		
+		/*
+		* Getting to be a bit edge-case here, lets remove
+		* the quoted strings from the statement, and 
+		* check again. We don't care that the sql makes no
+		* sense because we aren't going to use it
+		*/
+		foreach ($quotedStringArray[0] as $qString)
+			$tempSql = str_replace($qString,'',$tempSql);
+
+		return substr_count($tempSql,'?');
+		
+	}
+	/**
+	 * Execute SQL
 	 *
-	 * @param string|array $sql
-	 * @param array $inputarr
+	 * @param string     $sql      SQL statement to execute, or possibly an array
+	 *                             holding prepared statement ($sql[0] will hold sql text)
+	 * @param array|bool $inputarr holds the input data to bind to.
+	 *                             Null elements will be set to null.
 	 *
-	 * @return bool|mysqli_result
+	 * @return ADORecordSet|bool
 	 */
+	public function execute($sql, $inputarr = false) {
+		
+		if ($this->fnExecute) {
+			$fn = $this->fnExecute;
+			$ret = $fn($this,$sql,$inputarr);
+			if (isset($ret)) {
+				return $ret;
+			}
+		}
+		
+		if ($inputarr === false) {
+			return $this->_execute($sql,false);
+		}
+		
+		if (!is_array($inputarr)) {
+			$inputarr = array($inputarr);
+		}
+
+		if (!is_array($sql)) {
+			
+			/* 
+			* Make sure the number of parameters provided in the input
+			* array matches what the query expects
+			*/
+			$nparams = $this->placeHolderCount($sql);
+			
+			if ($nparams != count($inputarr)) {
+				$this->outp_throw(
+					"Input array has " . count($inputarr) .
+					" params, does not match query: '" . htmlspecialchars($sql) . "'",
+					'Execute'
+				);
+				return false;
+			}
+
+			$typeString = '';
+			$typeArray  = array(''); //placeholder for type list
+
+			foreach ($inputarr as $v) 
+			{
+				$typeArray[] = $v;
+				if (is_integer($v) || is_bool($v))
+					$typeString .= 'i';
+				
+				else if (is_float($v))
+					$typeString .= 'd';
+					
+				else if(is_object($v))
+					/*
+					* Assume a blob
+					*/
+					$typeString .= 'b';
+			
+				else
+					$typeString .= 's';
+				
+			} 
+				
+			/*
+			* Place the field type list at the front of the
+			* parameter array. This is the mysql specific
+			* format
+			*/
+			$typeArray[0] = $typeString;
+		
+			$ret = $this->_execute($sql,$typeArray);
+			if (!$ret) {
+				return $ret;
+			}
```

Comment:
@mnewnham I came upon this as I was reviewing your patch for #806.

As the code stands, this if statement is useless and could be removed, because $ret is already being returned at the end of the function, just a few lines later. Maybe you meant to `return false;` ?

Reply:
Mark's reply in https://github.com/ADOdb/ADOdb/pull/816#issuecomment-1076967876
> I did review that uneeded bit of code, you are correct in that it could be removed, so if its easiest you could just chop it out

Reply:
Fixed in 88aa5d0f75d9888b6b6ce60f75cfdeb35427b136